File: Service.Implementation/TemporaryDirectory.fs
namespace Service.Implementation.Utils

open System.IO
open System
open Service

type TemporaryDirectory() = 

    let path = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName())
    let dirInfo = Directory.CreateDirectory(path)

    member this.Path = path

    interface IDisposable with 
        member this.Dispose() = Directory.Delete(path, recursive=true)

    member this.FullPath fileName = Path.Combine(path, fileName)

    member this.Store (file:Dto.FileData) = 
        let filePath = this.FullPath file.Name
        File.WriteAllText(filePath, file.Contents)
        filePath

File: Service.Implementation/Utils/TemporaryDirectory.cs
using Service.Dto;
using System;
using System.IO;

namespace Service.Implementation.Utils
{
    class TemporaryDirectory : IDisposable
    {
        public string Path { get; }

        public TemporaryDirectory()
        {
            Path = System.IO.Path.Combine(System.IO.Path.GetTempPath(), System.IO.Path.GetRandomFileName());
            Directory.CreateDirectory(Path);
        }

        protected virtual void Dispose(bool disposing)
        {
            Directory.Delete(Path, recursive: true);
        }

        ~TemporaryDirectory()
        {
            Dispose(false);
        }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        public string FullPath(string fileName)
        {
            return System.IO.Path.Combine(Path, fileName);
        }

        public string Store(FileData file)
        {
            var filePath = FullPath(file.Name);
            File.WriteAllText(filePath, file.Contents);
            return filePath;
        }
    }
}
